### PR TITLE
Don't load ad if the tag height is not an AdSense whitelisted height

### PR DESCRIPTION
**Pull request by noranazmy**

### CODE REVIEW

File: extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js [hunk 1]
```diff
@@ -61,6 +61,7 @@ import {insertAnalyticsElement} from '../../../src/extension-analytics';
 import {
   getAdSenseAmpAutoAdsExpBranch,
 } from '../../../ads/google/adsense-amp-auto-ads';
+import {validateAdSenseAdHeight} from '../../../ads/google/adsense.js'
```
Comment (riklund) [inline extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js]: I don't think you need the js extension here.

Comment (riklund) [inline extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js]: Do we need this check here as well? (the validateData check is only in the other file)

File: ads/google/adsense.js [hunk 2]
```diff
@@ -17,6 +17,21 @@
 import {validateData} from '../../3p/3p';
 import {setStyles} from '../../src/style';
 import {camelCaseToDash} from '../../src/string';
+import {user} from '../../src/log';
+
+/** @const {Array<number>} */
```
Comment (riklund) [inline ads/google/adsense.js]: Perhaps no need for a list if there's only one value so far.

File: extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js [hunk 3]
```diff
@@ -212,6 +213,9 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
         isInManualExperiment(this.element);
     const width = Number(this.element.getAttribute('width'));
     const height = Number(this.element.getAttribute('height'));
+
+    validateAdSenseAdHeight(height);
```
Comment (riklund) [inline extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js]: Shouldn't we only do this check if it's a responsive ad?

File: ads/google/adsense.js [hunk 4]
```diff
@@ -17,6 +17,21 @@
...
+const WHITELISTED_HEIGHTS = [150];
+
+/**
+ * Check that the <amp-ad> tag height has not been tampered with.
```
Comment (riklund) [inline ads/google/adsense.js]: Nit: s/"has not been tampered with"/"is one of the allowed values for responsive ad units" (or similar)

File: ads/google/adsense.js [hunk 5]
```diff
@@ -17,6 +17,21 @@
...
+ * @param {Number} height
+ */
+export function validateAdSenseAdHeight(height) {
```
Comment (riklund) [inline ads/google/adsense.js]: validateResponsiveAdHeight (or similar)?

**Review by riklund — CHANGES_REQUESTED** [review]
Thanks Noran!

File: extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js (hunk 1 above)
Comment (noranazmy) [reply extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js]: Yes, this is necessary for the a4a path, doesn't work without it.

File: extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js (hunk 3 above)
Comment (noranazmy) [reply extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js]: I was thinking the same, will add the check.

File: ads/google/adsense.js (hunk 2 above)
Comment (noranazmy) [reply ads/google/adsense.js]: I was wondering if we will have different approved values or not. Seemed like a generic solution would be better in this case. Maybe @charliereams has something to say on that.

Comment (riklund) [reply ads/google/adsense.js]: I don't think I've heard of any such plans, but Charlie can correct me if I'm wrong here.

File: ads/google/adsense.js [hunk 6]
```diff
@@ -28,6 +44,9 @@ export function adsense(global, data) {
   validateData(data, [],
       ['adClient', 'adSlot', 'adHost', 'adtest', 'tagOrigin', 'experimentId',
         'ampSlotIndex', 'adChannel', 'autoFormat']);
+  if (data.autoFormat === 'rspv') {
```
Comment (riklund) [inline ads/google/adsense.js]: Actually, what do you think about putting this and the validateData call above in its own export?
That way we could implement the TODO in getAdUrl and not need to call it from two different places.

**Review by riklund — CHANGES_REQUESTED** [review]

**Comment by noranazmy** [comment]
@riklund, your comments should be addressed by now, what do you think?

File: ads/google/utils.js [hunk 7]
```diff
@@ -133,3 +134,31 @@ function validateDimensions(width, height, widthCond, heightCond, errorBuilder)
...
+/**
+ * Check that the <amp-ad> tag height is one of the allowed heights for
+ * responsive ad units.
+ * @param {Number} height
+ */
+export function validateResponsiveAdHeight(height) {
+  user().assert(
+      height === WHITELISTED_HEIGHT,
+      'Specified height ' + height +
+          ' in <amp-ad> tag is not one of the approved heights for AdSense.');
```
Comment (riklund) [inline ads/google/utils.js]: Nit: s/one of the approved heights/is not equal to the required value of WHITELISTED_HEIGHT for responsive AdSense ad units/ (or similar).

File: ads/google/utils.js [hunk 8]
```diff
@@ -133,3 +134,31 @@ function validateDimensions(width, height, widthCond, heightCond, errorBuilder)
...
+
+/** @const {Number} */
+const WHITELISTED_HEIGHT = 150;
+
+/**
+ * Check that the <amp-ad> tag height is one of the allowed heights for
+ * responsive ad units.
+ * @param {Number} height
+ */
+export function validateResponsiveAdHeight(height) {
```
Comment (riklund) [inline ads/google/utils.js]: Nit: No need to export this.

File: ads/google/utils.js [hunk 9]
```diff
@@ -133,3 +134,31 @@ function validateDimensions(width, height, widthCond, heightCond, errorBuilder)
   }
   return !badParams.length;
 }
+
+/** @const {Number} */
+const WHITELISTED_HEIGHT = 150;
```
Comment (riklund) [inline ads/google/utils.js]: Perhaps not for this PR, but just a note that we'll probably make this 320 in the end.

File: ads/google/utils.js [hunk 10]
```diff
@@ -133,3 +134,31 @@ function validateDimensions(width, height, widthCond, heightCond, errorBuilder)
...
+/**
+ * Validate <amp-ad> data and validate height for responsive ad units.
+ */
+export function validateAdSenseData(data) {
+  // TODO: check mandatory fields
+  validateData(data, [],
+      ['adClient', 'adSlot', 'adHost', 'adtest', 'tagOrigin', 'experimentId',
+        'ampSlotIndex', 'adChannel', 'autoFormat']);
+  if (data.autoFormat === 'rspv') {
+    validateResponsiveAdHeight(data.height);
```
Comment (riklund) [inline ads/google/utils.js]: I'm not 100% sure how adsense(global, data) gets called, but I strongly suspect that 'data' there is computed from element.dataset like in amphtml/src/3p-frame.js, so it probably won't have 'height' and 'width' in the general case.

File: extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js [hunk 11]
```diff
@@ -200,20 +201,35 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
...
+      'adtest' : this.element.getAttribute('data-adtest'),
+      'tagOrigin' : this.element.getAttribute('data-tag-origin'),
+      'experimentId' : this.element.getAttribute('data-experiment-id'),
+      'ampSlotIndex' : this.element.getAttribute('data-amp-slot-index'),
+      'adChannel' : this.element.getAttribute('data-ad-channel'),
+      'autoFormat' : this.element.getAttribute('data-auto-format'),
+      'width' : Number(this.element.getAttribute('width')),
+      'height' : Number(this.element.getAttribute('height'))
+    };
+    validateAdSenseData(data);
```
Comment (riklund) [inline extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js]: I don't think this'll do exactly the thing we want. data should probably originate from element.dataset (see previous comment) and also include any data attributes not specified in order for the validation to check that they aren't there.

**Review by riklund — COMMENTED** [review]
Thanks!

File: ads/google/utils.js (hunk 7 above)
Comment (noranazmy) [reply ads/google/utils.js]: Isn't this more of a user-friendly message than a technical message? Should we be exposing the variable name?

Comment (riklund) [reply ads/google/utils.js]: Sorry, I meant to expand the value of the variable, wasn't clear from my comment.

**Pull request closed without merging**